File: src/domain/provider.ts
import type { Availability, Bytes, EventKind, EventTiming, Participation } from "./calendar.ts";

export type ProviderKind = "google" | "microsoft" | "caldav" | "local";
export type ProviderConnectionState = "disconnected" | "authorizing" | "ready" | "error";
export type ProviderChangeKind = "upsert" | "delete";
export type ProviderMutationKind = "create" | "update" | "delete";

/** Safe account metadata. Credential material belongs in platform storage. */
export interface ProviderAccount {
  readonly id: number;
  readonly provider: ProviderKind;
  readonly displayName: Bytes;
  readonly state: ProviderConnectionState;
}

/** Opaque incremental-sync token; the provider adapter alone interprets it. */
export interface ProviderSyncCursor {
  readonly accountId: number;
  readonly value: Bytes;
}

/**
 * Provider adapter output before stable local ids are assigned. This is the
 * only vendor-facing event shape the domain ingestion path needs.
 */
export interface NormalizedProviderEvent {
  readonly accountId: number;
  readonly externalCalendarId: Bytes;
  readonly externalEventId: Bytes;
  readonly revision: Bytes;
  readonly title: Bytes;
  readonly location: Bytes;
  readonly startsAtMs: number;
  readonly endsAtMs: number;
  readonly timing: EventTiming;
  readonly eventKind: EventKind;
  readonly participation: Participation;
  readonly availability: Availability;
}

export interface ProviderChange {
  readonly kind: ProviderChangeKind;
  readonly event: NormalizedProviderEvent;
}

/**
 * Mutation envelope sent toward an adapter. The payload stays normalized;
 * each adapter translates it into the provider's request schema.
 */
export interface ProviderMutation {
  readonly id: number;
  readonly kind: ProviderMutationKind;
  readonly event: NormalizedProviderEvent;
}

File: src/domain/provider.zig
//! Typed contracts at the provider effect boundary.

const calendar = @import("calendar.zig");

pub const ProviderKind = enum { google, microsoft, caldav, local };
pub const ProviderConnectionState = enum { disconnected, authorizing, ready, error_state };
pub const ProviderChangeKind = enum { upsert, delete };
pub const ProviderMutationKind = enum { create, update, delete };

/// Safe account metadata. Credential material belongs in platform storage.
pub const ProviderAccount = struct {
    id: u64,
    provider: ProviderKind,
    display_name: []const u8,
    state: ProviderConnectionState,
};

/// Opaque incremental-sync token; the provider adapter alone interprets it.
pub const ProviderSyncCursor = struct {
    account_id: u64,
    value: []const u8,
};

/// Provider adapter output before stable local ids are assigned.
pub const NormalizedProviderEvent = struct {
    account_id: u64,
    external_calendar_id: []const u8,
    external_event_id: []const u8,
    revision: []const u8,
    title: []const u8,
    location: []const u8,
    starts_at_ms: i64,
    ends_at_ms: i64,
    timing: calendar.EventTiming,
    event_kind: calendar.EventKind,
    participation: calendar.Participation,
    availability: calendar.Availability,
};

pub const ProviderChange = struct {
    kind: ProviderChangeKind,
    event: NormalizedProviderEvent,
};

/// Mutation sent toward an adapter; each adapter owns vendor translation.
pub const ProviderMutation = struct {
    id: u64,
    kind: ProviderMutationKind,
    event: NormalizedProviderEvent,
};
